bootstrap/src/basis/file.ml

```diff
@@ -0,0 +1,243 @@
+  end
+
+let _seek_hlt strategy i t =
+  match _seek strategy i t with
+  | Error e -> halt e
+  | Ok offset -> offset
+
+let seek i t =
+  _seek Unix.SEEK_CUR i t
+
+let seek_hlt i t =
+  _seek_hlt Unix.SEEK_CUR i t
+
+let seek_left u t =
+  let i = isize_of_usize u in
+  _seek Unix.SEEK_SET i t
+
+let seek_left_hlt u t =
+  let i = isize_of_usize u in
+  _seek_hlt Unix.SEEK_SET i t
+
+let seek_right u t =
+  let i = isize_of_usize u in
+  _seek Unix.SEEK_END i t
+
+let seek_right_hlt u t =
+  let i = isize_of_usize u in
+  _seek_hlt Unix.SEEK_END i t
+
+module Stream = struct
+
+  type outer = t
+  type t = Byte.t Stream.t
+
+  let of_file file =
+    let rec fn i bytes file = lazy begin
+      match  i < Array.length bytes with
+      | false -> begin
+        let bytes_res = read file in
+        match bytes_res with
+        | Error _
+        | Ok [||] -> begin
+          let _ = close file in
+          Stream.Nil
+        end
+        | Ok bytes' -> Lazy.force (fn 0 bytes' file)
+      end
+      | true -> begin
+        let elm = Array.get i bytes in
+        let i' = succ i in
+        let byte_stream = fn i' bytes file in
+        Stream.Cons(elm, byte_stream)
+      end
+    end in
+    fn 0 [||] file
+
+  let of_path path =
+    let file_res = of_path ~mode:Mode.R_O path in
+    match file_res with
+    | Error error -> Error error
+    | Ok file -> Ok (of_file file)
+
+  let of_path_hlt path =
+    let t_res = of_path path in
+    match t_res with
+    | Error e -> halt e
+    | Ok(t) -> t
+
+  let write file t =
+    let rec fn file t = begin
+      match t with
+      | lazy Stream.Nil -> None
+      | lazy (Stream.Cons(_, _)) -> begin
+        let t', t'' = Stream.split buflen t in
+        let bytes = Array.of_stream t' in
+        let error_opt = write bytes file in
+        match error_opt with
+        | None -> fn file t''
+        | Some error -> Some error
+      end
+    end in
+    fn file t
+
+  let write_hlt file t =
+    match write file t with
+    | Some e -> halt e
+    | None -> ()
+
+end
+
+let%expect_test "of_path_hlt" =
+  let open Format in
+  printf "@[<h>";
+  let t = of_path_hlt ~mode:Mode.R_O "/home/cevans/test.txt" in
+  let bytes = read_hlt t in
+  let s = Bytes.to_string_hlt bytes in
+  let _ = close_hlt t in
+  printf "t = %a\n" String.pp s;
+  printf "@]";
+
+  [%expect{|
+    t = "The quick brown fox jumped over the red fence.\n"
+    |}]
+
+let%expect_test "Stream.of_path_hlt" =
+  let open Format in
+  printf "@[<h>";
+  Stream.of_path_hlt "/home/cevans/test.txt"
+  |> Array.of_stream
+  |> Bytes.to_string_hlt
+  |> printf "%a\n" String.pp;
+  printf "@]";
+
+  [%expect{|
+    "The quick brown fox jumped over the red fence.\n"
+    |}]
+
+let%expect_test "Stream.of_path_hlt" =
+  let open Format in
+  printf "@[<h>";
+  let _ = Stream.of_path_hlt "/home/cevans/test.txt"
+    |> Stream.write_hlt (of_path_hlt ~mode:Mode.W_A "/home/cevans/test2.txt")
+   in
+  printf "@]";
+
+  [%expect{| |}]
```

Comment:
I'll get rid of this.

Reply:
Note that it's possible to use the stand-alone test infrastructure similar to what I did for the entropy tests (in `bootstrap/test/basis/`) for the cases where you need a test file to operate on.